cou/steps/hypervisor.py

```diff
@@ -148,29 +150,51 @@ def azs(self) -> AZs:
 
         return azs
 
-    def _generate_pre_upgrade_plan(self) -> PreUpgradeStep:
+    def _generate_pre_upgrade_steps(
+        self, target: OpenStackRelease, group: HypervisorGroup
+    ) -> list[PreUpgradeStep]:
         """Generate pre upgrade plan for all applications.
 
-        This section should create a plan with steps like changing charm config option, etc.
+        This section should create a list of steps like changing charm config option, etc.
 
-        :return: Pre-upgrade step with all needed pre-upgrade steps.
-        :rtype: PreUpgradeStep
+        :param target: OpenStack codename to upgrade.
+        :type target: OpenStackRelease
+        :param group: HypervisorGroup object
+        :type group: HypervisorGroup
+        :return: List of pre-upgrade steps.
+        :rtype: list[PreUpgradeStep]
         """
-        raise NotImplementedError
+        steps = []
+        for app in self.apps:
+            if app.name not in group.app_units:
+                logger.debug(
+                    "skipping application %s because it is not part of group %s",
+                    app.name,
+                    group.name,
+                )
+                continue
 
-    def _generate_hypervisor_group_upgrade_plan(
-        self, target: OpenStackRelease, group: HypervisorGroup
-    ) -> HypervisorUpgradePlan:
+            units = group.app_units[app.name]
+            logger.info("generating pre-upgrade steps for %s app and %s units", app.name, units)
```

Comment:
The current phrase sounds like the app and units are not necessary related. I suggest updating it a little:
```suggestion
            logger.info("generating pre-upgrade steps for %s units of %s app", app.name, units)
```